src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java

```diff
@@ -154,31 +326,203 @@ private void registerMBean()
+     * probeTimerMillis members set.
+     */
+    @VisibleForTesting
+    static void calculateProbes(Map<InetAddressAndPort, AnnotatedMeasurement> samples, long intervalMillis) {
+        for (Map.Entry<InetAddressAndPort, AnnotatedMeasurement> entry: samples.entrySet())
+        {
+            if (entry.getKey().equals(FBUtilities.getBroadcastAddressAndPort()))
+                continue;
+
+            AnnotatedMeasurement measurement = entry.getValue();
+            long lastMeasure = measurement.millisSinceLastMeasure.getAndAdd(intervalMillis);
+            long lastRequest = measurement.millisSinceLastRequest.getAndAdd(intervalMillis);
+
+            if (lastMeasure >= MIN_PROBE_INTERVAL_MS && lastRequest < MAX_PROBE_INTERVAL_MS)
+            {
+                if (measurement.probeTimerMillis == 0)
+                {
+                    measurement.probeTimerMillis = intervalMillis;
+                }
+                else if (measurement.probeFuture != null && measurement.probeFuture.isDone())
+                {
+                    measurement.probeTimerMillis = Math.min(MAX_PROBE_INTERVAL_MS, measurement.probeTimerMillis * 2);
+                }
+            }
+            else
+            {
+                measurement.probeTimerMillis = 0;
+            }
+        }
+    }
+
+    @VisibleForTesting
+    void schedulePings(Map<InetAddressAndPort, AnnotatedMeasurement> samples)
+    {
+        for (Map.Entry<InetAddressAndPort, AnnotatedMeasurement> entry: samples.entrySet())
+        {
+            AnnotatedMeasurement measurement = entry.getValue();
+            long delay = measurement.probeTimerMillis;
+            long millisSinceLastRequest = measurement.millisSinceLastRequest.get();
+
+            if (millisSinceLastRequest > MAX_PROBE_INTERVAL_MS && !Gossiper.instance.isAlive(entry.getKey()))
+            {
+                samples.remove(entry.getKey());
+            }
+
+            if (delay > 0 && millisSinceLastRequest < MAX_PROBE_INTERVAL_MS &&
+                (measurement.probeFuture == null || measurement.probeFuture.isDone()) &&
+                !updateSamplesScheduler.isCancelled())
+            {
+                logger.trace("Scheduled latency probe against {} in {}ms", entry.getKey(), delay);
+                measurement.probeFuture = latencyProbeExecutor.schedule(() -> sendPingMessageToPeer(entry.getKey()),
+                                                                        delay, TimeUnit.MILLISECONDS);
+            }
+        }
+    }
+
+    /**
+     * Method that actually sends latency probes as PING messages. This is the only function in this class
+     * that operates on the latencyProbeExecutor thread and it records the maximum latency between a small and large
+     * message channel ping.
+     */
+    private void sendPingMessageToPeer(InetAddressAndPort to)
+    {
+        probeRateLimiter.acquire(dynamicLatencyProbeInterval);
+
+        if (!StorageService.instance.isGossipActive())
```

Comment:
Because this thread can run up to 10 minutes after it was scheduled, we maybe shouldn't be talking to other nodes at this junction (i.e. gossip is turned off). I know it's racey in the sense that we can check, then gossip can be turned off, then we can send a ping; but realistically users are going to be having issues where they're trying to drain the node and then still get messages 10 minutes later out of this.